packages/gatsby-plugin-gatsby-cloud/src/components/Indicator.js

```diff
@@ -35,6 +40,69 @@ const Indicator = () => {
   const trackedInitialLoad = useRef(false)
   const { track } = useTrackEvent()
 
+  async function fetchPageData() {
+    const urlHostString = window.location.origin
+    const pathAdjustment =
+      window.location.pathname === `/` ? `/index` : window.location.pathname
+
+    const url = `${urlHostString}/page-data${pathAdjustment}/page-data.json`
+    try {
+      const resp = await fetch(url)
+      const data = await resp.text()
+
+      if (data === `{}`) {
+        // for local development, force and error if page is missing.
+        const err = new Error(`Not Found`)
+        err.status = 404
+        throw err
+      }
+
+      return { data, errorMessage: null }
+    } catch (e) {
+      if (e.status === 404) {
+        return { data: null, errorMessage: `This page has moved.` }
+      } else {
+        return { data: null, errorMessage: null }
+      }
+    }
+  }
+
+  useEffect(() => {
+    const fetchData = async () => {
+      const { data } = await fetchPageData()
+      pageData = data
+    }
+
+    fetchData()
+  }, [])
+
+  const hasPageDataChanged = async () => {
+    if (buildId !== latestCheckedBuild) {
+      let pageDataCounter = 0
+      let hasPageChanged = false
+
+      while (!hasPageChanged && pageDataCounter <= PAGE_DATA_RETRY_LIMIT) {
+        const loadedPageData = pageData
+        const { data, errorMessage } = await fetchPageData()
+
+        if (errorMessage) {
+          return { hasPageChanged: false, errorMessage }
+        }
+
+        pageDataCounter++
+        hasPageChanged = loadedPageData !== data
```

Comment:
This could be problematic since we are just checking pointer equality here. Since `fetchPageData` always returns a new object, these two objects will always evaluate to "not equal", even if the contents of the objects is the same.

As a solution, you could do some sort of deep equality check here. Or if that is getting too expensive, we could get clever with a different pointer comparison solution. I'm glad to help you figure that out if you want to pair on this!

Reply:
if I understand correctly, `fetchPageData` returns a string because of the `Response.text()` call. I agree it's a little confusing because right after we check to see if `data === {}` which makes it look like we're expecting `data` to always be an `{}`. So maybe instead of checking if `data === {}` locally, do the change that Matt had mentioned with a NODE_ENV check instead.

Reply:
Aha! Yup, my comment was assuming that `fetchPageData` did not return a string, but since it does return a string, this should be fine as is :ok_hand: